### PR TITLE
proxy missing handler/target marshalling

### PR DESCRIPTION
**Pull request by jianchun**
Fix a list of proxy bugs related to Var marshalling.

* handler/target fields have no guarantee if they match caller
  requestContext. Should marshal them explicitly when calling traps.

* 5 traps (delete, getPrototypeOf, has, isExtensible, preventExtensions)
  were incorrectly using proxy context as requestContext because their
  callers do not provide requestContexts. Changing callers would be
  very noisy. Fixed them instead by retrieving requestContext from host
  scriptContext stack.

* 1 trap (setPrototypeOf) had the same problem but adding a requestContext
  is trivial. Fixed by adding requestContext in caller.

* 1 trap (ownKeys) had requestContext but still incorrectly mixed with
  proxy context. Fixed this and others to use requestContext.

* Deleted "proxy_cctx_bugs.js" (dangling file due to merge error).

Other changes:

* Move implicit call checks to top consistenly.

* Added implicit exception check consistently.

* Fixed a typo in GetPropertyDescriptorTrap error message (
  "preventExtensions" -> "getOwnPropertyDescriptor").

* Removed an aggressive assertion in ScriptFunction::ChangeEntryPoint. The
  original assertion was to verify against "checkCodeGenThunk". Later NoJIT
  refactoring added the other over aggressive branch. The script function
  can be a CrossSiteObject, e.g. when a CrossSite script function is
  defer-parsed.

### CODE REVIEW

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 1]
```diff
@@ -1161,15 +1173,19 @@ namespace Js
...
         // Reject implicit call
-        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        ThreadContext* threadContext = GetScriptContext()->GetThreadContext();
         if (threadContext->IsDisableImplicitCall())
         {
             threadContext->AddImplicitCallFlags(Js::ImplicitCall_External);
             return  FALSE;
         }
 
+        // JavascriptOperators::DeleteProperty does not pass down requestContext.
```
Comment (kunalspathak) [inline lib/Runtime/Library/JavascriptProxy.cpp]: >DeleteProperty [](start = 32, length = 14)

nit:PreventExtensions?

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 2]
```diff
@@ -1098,14 +1105,19 @@ namespace Js
...
         // Reject implicit call
-        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        ThreadContext* threadContext = GetScriptContext()->GetThreadContext();
         if (threadContext->IsDisableImplicitCall())
         {
             threadContext->AddImplicitCallFlags(Js::ImplicitCall_External);
             return FALSE;
         }
+
+        // JavascriptOperators::DeleteProperty does not pass down requestContext.
```
Comment (kunalspathak) [inline lib/Runtime/Library/JavascriptProxy.cpp]: >DeleteProperty [](start = 32, length = 14)

nit: isextensible?

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 3]
```diff
@@ -1432,15 +1448,19 @@ namespace Js
...
-        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        ThreadContext* threadContext = GetScriptContext()->GetThreadContext();
         if (threadContext->IsDisableImplicitCall())
         {
             threadContext->AddImplicitCallFlags(Js::ImplicitCall_External);
-            return scriptContext->GetLibrary()->GetUndefined();
+            return GetScriptContext()->GetLibrary()->GetUndefined();
         }
 
+        // JavascriptOperators::DeleteProperty does not pass down requestContext.
```
Comment (kunalspathak) [inline lib/Runtime/Library/JavascriptProxy.cpp]: >DeleteProperty [](start = 32, length = 14)

fix this.

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 4]
```diff
@@ -1719,8 +1739,8 @@ namespace Js
         CallInfo callInfo(CallFlags_Value, 4);
         Var varArgs[4];
         Js::Arguments arguments(callInfo, varArgs);
-        varArgs[0] = handlerObj;
-        varArgs[1] = targetObj;
+        varArgs[0] = proxy->GetHandlerArg(scriptContext);
```
Comment (kunalspathak) [inline lib/Runtime/Library/JavascriptProxy.cpp]: >scriptContext [](start = 42, length = 13)

can you rename this to `requestContext` to be in sync with other methods?

**Comment by kunalspathak** [comment]
>         varArgs[1] = targetObj;

do we need to fix here?

---
Refers to: lib/Runtime/Library/JavascriptProxy.cpp:2192 in 9aed10d. [](commit_id = 9aed10d92715d9dec34077e6883eadc56cc6d2e4, deletion_comment = False)

**Comment by jianchun** [comment]
> do we need to fix here?

(Mentioned in unit test) When cross-site this one is behind "CrossSiteProxyCallTrap" which ensures this is called in proxy context. So I did not touch this one.

**Comment by akroshg** [comment]
We do not need to marshal this one.

---
In reply to: [313195599](https://github.com/Microsoft/ChakraCore/pull/3286#issuecomment-313195599) [](ancestors = 313195599)

**Comment by akroshg** [comment]
:shipit:

**Comment by jianchun** [comment]
Thanks @kunalspathak @akroshg

**Pull request merged**